prime-router/src/main/kotlin/azure/ActionHistory.kt

```diff
@@ -0,0 +1,331 @@
+            OffsetDateTime nextActionAt,
+            String         sendingOrg,
+            String         sendingOrgClient,
+            String         receivingOrg,
+            String         receivingOrgSvc,
+            String         schemaName,
+            String         schemaTopic,
+            String         bodyUrl,
+            String         external_name,
+            String         bodyFormat,
+            byte[]         blobDigest,
+            Integer        itemCount,
+            OffsetDateTime wipedAt,
+            OffsetDateTime createdAt
+        */
+    /**
+     * Use this to record history info about an internally created report.
+     */
+    fun trackCreatedReport(
+        event: Event,
+        report: Report,
+        service: OrganizationService
+    ) {
+        if (isReportAlreadyTracked(report.id)) {
+            error("Bug:  attempt to track history of a report ($report.id) we've already associated with this action")
+        }
+
+        val reportFile = ReportFile()
+        reportFile.reportId = report.id
+        reportFile.nextAction = event.eventAction.toTaskAction()
+        reportFile.nextActionAt = event.at
+        reportFile.receivingOrg = service.organization.name
+        reportFile.receivingOrgSvc = service.name
+        reportFile.schemaName = report.schema.name
+        reportFile.schemaTopic = report.schema.topic
+        reportFile.bodyUrl = report.bodyURL
+        reportFile.bodyFormat = report.getBodyFormat().toString()
+        reportFile.itemCount = report.itemCount
+        reportsOut[reportFile.reportId] = reportFile
+    }
+
+    fun trackSentReport(
+        service: OrganizationService,
+        sentReportId: ReportId,
+        fileName: String?,
+        params: String,
+        result: String,
+        itemCount: Int
+    ) {
+        if (isReportAlreadyTracked(sentReportId)) {
+            error("Bug:  attempt to track history of a report ($sentReportId) we've already associated with this action")
+        }
+        val reportFile = ReportFile()
+        reportFile.reportId = sentReportId
+        reportFile.receivingOrg = service.organization.name
+        reportFile.receivingOrgSvc = service.name
+        reportFile.schemaName = service.schema
+        reportFile.schemaTopic = service.topic
+        reportFile.externalName = fileName
+        reportFile.transportParams = params
+        reportFile.transportResult = result
+        reportFile.bodyUrl = null
+        reportFile.bodyFormat = service.format.toString()
+        reportFile.itemCount = itemCount
+        reportsOut[reportFile.reportId] = reportFile
+    }
+
+    fun trackFailedReport(service: OrganizationService, sentReportId: ReportId, params: String, msg: String) {
+        trackSentReport(service, sentReportId, null, params, msg, 0)
+    }
+
+    /**
+     * Save the history about this action and related reports
+     */
+    fun saveToDb(txn: Configuration) {
+        insertAll(txn)
+    }
+
```

Comment:
I'm wondering if a fluent style may be a better choice for this object. Create a builder and then load to db.